### PR TITLE
[fixes #22] Introduce FastArray, a frozen Array with fast inclusion lookup

### LINKED ISSUE #22: Type lists should be sets
E.g. `TRUTHY_LITERALS`

Although there's not much valid reason to use them directly, this could break some code relying on them, e.g. `[...] - Set[...]` TypeError (no implicit conversion of Set into Array), so v1 might be a good time...

### PR DESCRIPTION
**Pull request by marcandre**
Creates an Array-like class with `===` and `include?` from Set.

### CODE REVIEW

File: lib/rubocop/ast/tuple.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module AST
+    # A frozen Array/Set
+    #
+    class Tuple < ::Array
```
[the remark this answers is not in the document]
Comment (koic) [reply lib/rubocop/ast/tuple.rb:7]: I've seen the concept of Tuple in Rinda, but I'm not familiar with it 😅 
https://ruby-doc.org/stdlib-2.7.1/libdoc/rinda/rdoc/Rinda/Tuple.html

Array, Hash and Set do not inherit from each other. I'm wondering if Tuple is a concept related to `is_a` of Array or not. The following is an example of Rinda:

```
% ruby -rrinda/rinda -e 'p Rinda::Tuple.ancestors'
[Rinda::Tuple, Object, Kernel, BasicObject]
```

[the remark this answers is not in the document]
Comment (bquorning) [reply lib/rubocop/ast/tuple.rb:7]: How come we cannot use a `Set`?

[the remark this answers is not in the document]
Comment (marcandre) [reply lib/rubocop/ast/tuple.rb:7]: @koic To be honest, I'm not completely satisfied with `Tuple`, a more descriptive name would be more like `DeeplyFrozenIndexedArray`. And that `is_a? Array`...

@bquorning That was [the first thing I tried](https://github.com/rubocop-hq/rubocop-ast/pull/26), but `Set` are [sorely lacking](https://gist.github.com/marcandre/1cf2f78e58977381dbdb3374526ef266) and the first two points created annoying issues. The choice was to have potential incompatibilities and have to worry about which "lists" were Sets (most) or Arrays (some that didn't need indexing and used `join`, say).

[the remark this answers is not in the document]
Comment (Drenmi) [reply lib/rubocop/ast/tuple.rb:17]: > To be honest, I'm not completely satisfied with `Tuple`, a more descriptive name would be more like `DeeplyFrozenIndexedArray`. And that `is_a? Array`...

I was thinking if `FastArray` could be appropriate. It gives a hint as to why we introduced this, and it also honours the fact that it is inherited from `Array`. 🤔

[the remark this answers is not in the document]
Comment (bbatsov) [reply lib/rubocop/ast/tuple.rb:17]: Yeah, I agree we can come with a better name than Tuple. Btw, I'm also thinking we might bend the rules and monkey patch Array to limit the prominence of the constructors in the codebase (e.g. `to_dfia` :D). Might be a good opportunity to use refinements for once, although I'm not sure if they have any negative performance implications.

[the remark this answers is not in the document]
Comment (Drenmi) [reply lib/rubocop/ast/tuple.rb:17]: Yes. Refinements are a fine (no pun intended) feature that might fit well here. Also have never tested the performance implications of them (but use them actively at work.) 🤔

[the remark this answers is not in the document]
Comment (marcandre) [reply lib/rubocop/ast/tuple.rb:17]: Sure, `FastArray`, `ConstArray`, `SetArray` (as in "fixed" and as "set-like" :-) ),`IndexedArray`

[the remark this answers is not in the document]
Comment (marcandre) [reply lib/rubocop/ast/tuple.rb:17]: So, is there a naming preference? `Tuple` / `FastArray` / `IndexedArray` / <something else>?

[the remark this answers is not in the document]
Comment (bbatsov) [reply lib/rubocop/ast/tuple.rb:17]: Well, arrays are supposed to be indexed by default, so I don't like `IndexedArray`. Seems to me that `FastArray` is the best out of the proposed names.

[the remark this answers is not in the document]
Comment (marcandre) [reply lib/rubocop/ast/tuple.rb]: Replaced with `FastArray` & rebased

File: lib/rubocop/ast/fast_array.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      def include?(value)
+        @to_set.include?(value)
+      end
+
+      alias === include?
+    end
+  end
+end
+
+def FastArray(list) # rubocop:disable Naming/MethodName
```
Comment (bquorning) [inline lib/rubocop/ast/fast_array.rb:49]: I am not too crazy about us defining constants outside of the `RuboCop` namespace. Is it necessary? As far as I can see, we only use `FastArray` _inside_ the `RuboCop::AST` namespace.

Comment (bbatsov) [reply lib/rubocop/ast/fast_array.rb:49]: I still think it's going to be better to define some refinement for use within RuboCop's code base and just use a conversion method like `.to_fa` everywhere.

Comment (marcandre) [reply lib/rubocop/ast/fast_array.rb:49]: The constant `FastArray` is defined within `RuboCop`, but I also added a global method `FastArray`, but it isn't strictly necessary, I could use `FastArray[ary]`. I can try with refinements too. Is `to_fast_array` ok? I fear `to_fa` would be quite cryptic

Comment (bquorning) [reply lib/rubocop/ast/fast_array.rb:49]: In my personal opinion, if we could get away with defining the `FastArray` method inside `RuboCop` scope, that would be optimal.

Comment (marcandre) [reply lib/rubocop/ast/fast_array.rb:49]: > In my personal opinion, if we could get away with defining the `FastArray` method inside `RuboCop` scope, that would be optimal.

We could, but we'll need an explicit `extend FastArray::Method` or similar (not needed when inheriting from a class that already extend it, like `Cop::Base`). Let me try that first.

Comment (marcandre) [reply lib/rubocop/ast/fast_array.rb]: Force pushed without a global function, and on the rubocop side too.

**Comment by bbatsov** [comment]
I'm fine with this approach, but if the others have concerns I won't push for it. We can always revisit this down the road.

**Comment by marcandre** [comment]
Cool.

@bquorning did you have a chance to look at the approach without global function?

**Comment by bquorning** [comment]
Sorry, I forgot to check before now.

I really like the explicit `extend FastArray::Function`. It’s not as easy to use as having the function in global namespace, but there’s less confusion about where the function really comes from now.

**Comment by koic** [comment]
I have the following @bquorning opinion again.

> It’s a tradeoff between speed, readability and understandability.

https://github.com/rubocop-hq/rubocop/pull/8133#issuecomment-642880710

Performance improvements are welcome, however I love the ease and readability of Ruby itself.
I think merging this PR means applying `FastArray` to the entire RuboCop https://github.com/rubocop-hq/rubocop/pull/8133, and RuboCop party gem and beyond.
My wish is to be able to read and write with plain Array literal provided by Ruby. I would like to take up ease and readability of Ruby itself rather than about 5% performance improvement.

**Comment by marcandre** [comment]
> We can always revisit this down the road.

I just want to mention that one issue is that these constants are public. Most are arrays, some are sets, which is confusing and error prone. It's less of an issue to change them now (pre 1.0) than after.

**Comment by jonatas** [comment]
I'm not an avid contributor but I'd like to give an idea here:

Instead of enforcing the FastArray inline, we can do it via metaprogramming just patching the `class.constants` no?

```ruby
node_classes = RuboCop::AST.constants.grep(/Node$/).map(&RuboCop::AST.method(:const_get))
node_classes.each do |clazz|
  clazz.constants.each do |const|
    value = clazz.const_get(const)
    next unless value.grep(Symbol) != value # all should be symbols
    clazz.remove_const(const)
    clazz.const_set(const, FastArray(value))
  end 
 end
```

**Comment by marcandre** [comment]
@jonatas I like your spirit 😅

While most such constants are in `RuboCop::AST::Node` and such, there are other constants in the main gem's cops that would benefit from that treatment too. Still, it could be done with an opt-in, say `extend OptimizeArrayConst` or something.

Did you have a look at #57?

**Comment by jonatas** [comment]
I understood your point. Very nice idea on #57, I checked now :)
I did some benchmarking locally. This looks quite challenging 🥇 
 
@marcandre Have you shared the issue with the ruby-core team? because as far as I can see here, this looks like bring benefit to any ruby array lookup right?

maybe someone has some ideas about how to make it better for the array in general or introduce the FastArray in the core library as we have BasicObject that is lighter than Object.

**Comment by marcandre** [comment]
> this looks like bring benefit to any ruby array lookup right?

Not really, no. Arrays may have duplicated elements, are not meant for lookups, but are good for fast creation / insertions / accessing `n`-th element, low memory, etc. Sets may not have duplicated elements, and are meant for lookups and are slower for creation / insertions. For all these frozen constants, we really should be using `Sets`, but there are a few annoyances which I hope to get ruby-core to reconsider in https://bugs.ruby-lang.org/issues/16989

**Pull request closed without merging**